File: backend/multiplexer.py
"""세션을 무엇이 붙잡아 두는가 — **tmux 가 기본이고, 안 붙잡는 것도 선택이다.**

이 앱은 tmux 위에 지어졌다. 로컬 pane 도 원격 pane 도 tmux 세션에 attach 하므로 연결이
끊겨도, 백엔드가 재시작돼도 셸은 살아 있다. 그 위에 한 가지 선택만 더 있다:

- `TMUX`: 세션을 tmux 가 붙잡는다. 이 저장소의 기본이자 사실상의 전부다.
- `NONE`: 붙잡지 않는다. 셸 하나가 뜨고 탭을 닫으면 끝난다. tmux 가 없는 호스트나
  일회성 셸에 쓴다. 고장이 아니라 유효한 선택이다.

한때 세 번째 값(herdr)이 있었다. 이 앱의 컨셉을 "tmux 기반 터미널" 로 다시 잡으면서
통째로 걷어냈다 — 두 멀티플렉서를 나란히 두면 "살아있는 세션 목록" 을 두 곳에 묻고
설정과 현실이 어긋나는 자리가 두 배가 된다.

**대신 반드시 말한다.** 세션이 안 남는다는 것은 사용자가 **닫기 전에** 알아야 하는
사실이다. `persists()` 가 그 판정 하나이고, 화면은 그것만 본다.

⚠️ **`use_remote_tmux` 는 아직 살아 있다.** 예전 호스트 행과 예전 클라이언트가 보내는
값이라, 새 `multiplexer` 칸이 비어 있으면 그것으로 되짚는다(`from_host_row`). 두 값이
따로 놀지 않게 **읽는 곳은 여기 하나**로 모은다.
"""
from __future__ import annotations

TMUX = "tmux"
NONE = "none"

#: 화면의 선택지 순서이기도 하다.
CHOICES: tuple[str, ...] = (TMUX, NONE)

#: 아무 것도 안 고른 호스트의 기본. 이 저장소가 tmux 위에 지어졌으므로 tmux 다.
DEFAULT = TMUX


def normalize(value: object, *, fallback: str = DEFAULT) -> str:
    """모르는 값은 조용히 fallback. 경계에서 한 번만 거른다.

    걷어낸 값(herdr)을 저장해 둔 옛 설정·호스트 행도 여기서 기본값으로 접힌다 —
    그 사용자의 세션이 죽은 것으로 읽히는 것보다 tmux 로 여는 쪽이 안전하다.
    """
    if isinstance(value, str):
        choice = value.strip().lower()
        if choice in CHOICES:
            return choice
    return fallback


def persists(choice: str) -> bool:
    """이 선택에서 세션이 접속을 넘어 살아남는가.

    화면의 "닫으면 끊어집니다" 안내는 이 한 줄만 본다 — 판정을 두 군데서 하면
    반드시 어긋난다(이 저장소가 탭 색·sanitize 에서 이미 두 번 밟았다).
    """
    return normalize(choice) != NONE


def from_host_row(row: dict | None, *, fallback: str = DEFAULT) -> str:
    """호스트 행 → 선택. 새 칸이 없으면 옛 `use_remote_tmux` 로 되짚는다.

    되짚기는 **끄기만** 할 수 있다: `use_remote_tmux=1` 은 "tmux" 라기보다
    "기본값 그대로" 라는 뜻이다.
    """
    if not row:
        return normalize(None, fallback=fallback)
    explicit = row.get("multiplexer")
    if explicit:
        return normalize(explicit, fallback=fallback)
    legacy = row.get("use_remote_tmux", 1)
    if legacy in (0, "0", False):
        return NONE
    return normalize(None, fallback=fallback)

File: frontend/src/utils/multiplexer.js
/**
 * 세션을 무엇이 붙잡아 두는가 — **tmux 가 기본이고, 안 붙잡는 것도 선택이다.**
 *
 * `backend/multiplexer.py` 의 거울이다. 두 쪽이 어긋나면 화면이 그리는 것과 서버가
 * 띄우는 것이 달라지는데, 그건 **저장하고 다시 연결해 보기 전까지 안 드러난다.**
 * 그래서 되짚기 규칙(`fromHost`)까지 여기 함께 둔다.
 *
 * ⚠️ **`none` 은 고장이 아니다.** 셸이 하나 뜨고 탭을 닫으면 끝난다 — 많은 쓰임에 그게
 * 맞다. 다만 그 사실은 **닫기 전에** 알려야 하고, 그 판정이 `persists()` 하나다.
 *
 * ⚠️ **고르는 자리는 설정 한 곳뿐이다**(설정 → 세션 멀티플렉서). 이 서버의 pane 도
 * 호스트의 pane 도 같은 값을 따른다. 호스트마다 또 고르게 두면 같은 결정이 두 자리에
 * 생기고, 전역 값을 바꿔도 옛 호스트가 따라오지 않는다. 호스트 행의 값은 옛
 * `use_remote_tmux=0` 을 존중하기 위해서만 읽는다.
 *
 * 한때 세 번째 값(herdr)이 있었고 통째로 걷어냈다. 저장돼 있던 그 값은 `normalize` 가
 * 기본값(tmux)으로 접는다.
 */

export const TMUX = 'tmux';
export const NONE = 'none';

/** 화면의 선택지 순서이기도 하다. */
export const CHOICES = [TMUX, NONE];

/** 아무 것도 안 고른 호스트의 기본. 이 저장소가 tmux 위에 지어졌으므로 tmux 다. */
export const DEFAULT = TMUX;

export const normalize = (value, fallback = DEFAULT) => (
  CHOICES.includes(value) ? value : fallback
);

/** 이 선택에서 세션이 접속을 넘어 살아남는가. 안내는 이 한 줄만 본다. */
export const persists = (choice) => normalize(choice) !== NONE;

/**
 * 호스트 행 → 선택. 새 칸이 없으면 옛 `use_remote_tmux` 로 되짚는다.
 *
 * 되짚기는 **끄기만** 표현할 수 있다: `use_remote_tmux=1` 은 "기본값 그대로" 라는 뜻이다.
 */
export const fromHost = (host, fallback = DEFAULT) => {
  if (!host) return normalize(null, fallback);
  if (host.multiplexer) return normalize(host.multiplexer, fallback);
  if (!host.use_remote_tmux) return NONE;
  return normalize(null, fallback);
};

/** 설정 셀렉트의 라벨 — 도구 이름은 번역하지 않는다. */
export const OPTIONS = [
  { value: TMUX, label: 'tmux' },
  { value: NONE, label: 'none' },
];

/** 고른 값이 실제로 무엇을 뜻하는지 한 줄. 특히 `none` 은 반드시 읽혀야 한다. */
export const HINTS = {
  [TMUX]: (t) => t?.('muxHintTmux')
    || '연결이 끊겨도 원격 tmux 가 세션을 살려둡니다.',
  [NONE]: (t) => t?.('muxHintNone')
    || '평범한 SSH 셸입니다. 탭을 닫거나 연결이 끊기면 실행 중이던 작업도 끝납니다.',
};
